Add ANARI Rendering Method to Volume Plot

griffin28 opened the pull request:
### Description

Resolves #20513 

<!-- Please include a summary of the change and any other information and instructions to help reviewers understand the work -->

<!-- Note that all the checklist items in various bulleted lists below end with ~~ characters.
     This is to facilitate striking them out when they do not apply.
     One just has to add ~~ characters just before the opening square bracket [ -->

### Type of change

<!-- Please check one of the boxes below -->

* ~~[ ] Bug fix~~
* [X] New feature
* [X] Documentation update
* ~~[ ] Other~~ <!-- please explain with a note below -->

### How Has This Been Tested?

1. Open `noise.silo`
2. Add _Volume -> chromeVf_ (or any variable of your choosing)
3. Select `ANARI` as the **Rendering Method**
4. Enable **ANARI Rendering**
5. Click Apply

### Checklist:

<!-- For items in this checklist that do not apply, simply insert two tilde chars, `~~`, just ahead of the left bracket char, `[` at the beginning of a line. Each line ends with two tilde chars to make doing such ~~strikeouts~~ easy. -->

- [X] I have commented my code where applicable.
- [X] I have updated the release notes.
- [X] I have made corresponding changes to the documentation.
- [X] I have added debugging support to my changes.
- ~~[ ] I have added tests that prove my fix is effective or that my feature works.~~
- [X] I have confirmed new and existing unit tests pass locally with my changes.
- ~~[ ] I have added new baselines for any new tests to the repo.~~
- ~~[ ] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.~~

[1]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/StyleGuide.html
[2]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers
[3]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/RCDevelopment.html#communication-protocols-and-public-apis

griffin28 committed:
Initial pass at adding ANARI volume rendering

griffin28 committed:
Initial pass at adding ANARI volume rendering

griffin28 committed:
Merge branch 'griffin28/anari/volume' of github.com:visit-dav/visit into griffin28/anari/volume

griffin28 committed:
ANARI volume rendering with panning, zooming, and removed useAnari

griffin28 committed:
Created ANARI rendering method for volume rendering

griffin28 committed:
Updated Anari enabled logic

griffin28 committed:
Added vtk anari patches to support volume rendering

griffin28 committed:
Merge branch 'develop' of github.com:visit-dav/visit into griffin28/anari/volume

griffin28 committed:
Docs, release notes updates, and final volume rendering edits

griffin28 committed:
Merge branch 'develop' of github.com:visit-dav/visit into griffin28/anari/volume

griffin28 committed:
Reverted experimental changes I didn't intend to commit

griffin28 committed:
Missed a couple of things on the revert

biagas commented on src/plots/Volume/anari/AnariVolumeWidget.h:
```diff
@@ -0,0 +1,117 @@
+// Copyright (c) Lawrence Livermore National Security, LLC and other VisIt
+// Project developers.  See the top-level LICENSE file for dates and other
+// details.  No copyright assignment is required to contribute to VisIt.
```
Just wondering if this is the same as the widget used in the gui's Rendering window, and if so, could we reduce code duplication by using the gui's widget directly?

griffin28 replied:
Unfortunately it's not. There are volume rendering specific things here while a lot of the code is the same.

biagas commented on src/plots/Volume/Volume.xml:
```diff
@@ -230,6 +242,23 @@
         0.000000
         15.000000
       </Field>
+      <Field name="anariRendering" label="ANARI Rendering" type="bool">
```
These seem a duplicate of the same fields in RenderingAttributes. It would be better to place the Anari options in their own state file, and have the RenderingAttributes and VolumeAttributes use their own versions.  This would reduce code duplication and make Anari easier to maintain/update in the future.

My antialiasing branch has an example where I created FXAAOptions state object, and added it to RenderingAttributes, you can take a look there for how to make this work:

https://github.com/visit-dav/visit/tree/biagas/make_antialiasing_multimodal

If you'd rather get this PR in, and make a new ticket to make this suggested change, I am ok with that, but it should be done for 3.5.0.

griffin28 replied:
Sounds good. I'll make a new ticket with the suggested change and make sure it's done for 3.5.0

biagas commented on src/plots/Volume/VolumeAttributes.code:
```diff
@@ -653,6 +651,3 @@ VolumeAttributes_SetMaterialProperties(PyObject *self, PyObject *args)
     return Py_None;
 }
 
-Target: xml2cmake
```
Why was this removed?

griffin28 replied:
Not sure...a bug in the generate code script? I used `xmledit` and selected generate code when adding the ANARI volume attributes. I can revert the change.

biagas replied:
Sorry about the comment, yes it is a bug in the xmledit writing to code file that I had forgotten about (#20529). Not a big deal as long as the xml2xxx tools still write the correct atts

biagas commented on src/doc/using_visit/Plots/PlotTypes/VolumePlot.rst:
```diff
@@ -54,6 +54,10 @@ Each sample point along the ray is composited to form a final color for the scre
 Rays are traced from closest to farthest to allow for early ray termination which stops the sampling process when the pixel opacity gets above a certain threshold.
 This method of volume-rendering yields superior pictures at the cost of speed and memory use.
 
+If VisIt_ is built with `ANARI <https://www.khronos.org/api/index_2017/anari/>`_ support by using the ``-–anari`` option, 
```
The style standard for our RST is one sentence per line. Would you please fix for the lines you have added.

griffin28 committed:
Fixed RST and reverted Target removal per PR feedback

Update baselines with minor pixel/shading diffs. (#20556)

Update avtXRayImageQuery.C (#20555)

Fix bug with cli not exiting when Quit initiated from GUI. (#20503)

Resolves #19432.

Stifle vtkLogger output. (#20498)

Resolves #20014.

Rebaseline plots/surface, unit/launcher, rendering/pixeldata. (#20560)

* Rebaseline plogs/surface, rendering/pixeldata, unit/launcher.

* Remove pixeldata cases from serial skip list.
Remove duplicate pixeldata listing from parallel skip list.

biagas commented:
@griffin28 Looks like you have a number of files now listed as changed that aren't part of your PR (some of them are things I merge with develop recently).  Did an update from develop go awry?

griffin28 committed:
Merge branch 'develop' of github.com:visit-dav/visit into griffin28/anari/volume

griffin28 committed:
Merge branch 'griffin28/anari/volume' of github.com:visit-dav/visit into griffin28/anari/volume

griffin28 commented:
Yeah...updated Ubuntu to 24 and some of my global git configs got overwritten. `git config pull.rebase`  was set to `true` so when I did a pull from develop it rebased my branch. Thought the squash would hide your commits but I was wrong.  So I reset to the last hash you checked and did a git pull with the correct setting (merge). Everything should be go to go now.

biagas commented on src/doc/using_visit/Plots/PlotTypes/VolumePlot.rst:
```diff
@@ -140,6 +144,52 @@ Lighting is enabled by default but the user can disable it by unchecking the **L
+    Because ANARI abstracts away the details of an entire rendering system, the underlying hardware which a back-end may use is entirely up to the implementation.
+    Please read your vendor's back-end documentation to see what parameters are available to configure and what underlying hardware is both available and used to render frames.
+
+``Back-end Subtype``: This option allows you to specify the subtype of the back-end to use if supported by the back-end. The **default** option is most common and supported by all of the back-ends.
```
```suggestion
``Back-end Subtype``: This option allows you to specify the subtype of the back-end to use if supported by the back-end.
The **default** option is most common and supported by all of the back-ends.
```

griffin28 replied:
Oppps...thought I got them all!

biagas replied:
No worries, I've been updating an .rst file myself, so my eyes are currently in tune to look for this!

biagas commented on src/plots/Volume/VolumeAttributes.code:
```diff
@@ -654,5 +652,3 @@ VolumeAttributes_SetMaterialProperties(PyObject *self, PyObject *args)
 }
 
 Target: xml2cmake
-Condition: VISIT_SLIVR
```
These two lines should have been added back in with the `Target: xml2cmake` line above. I apologize if I wasn't clear in my previous review and subseqent comments.

griffin28 replied:
Ok...I should've clarified before updating.

biagas reviewed (changes requested):
Minor fixes.

griffin28 committed:
Fixes from PR feedback

biagas commented on src/plots/Volume/avtVisItVTKRenderer.C:
```diff
@@ -640,12 +644,69 @@ avtVisItVTKRenderer::UpdateRenderingState(vtkDataSet * in_ds,
         }
         else
 #endif
+    }
```
Looks like the `else` on line 645 should be removed.

griffin28 replied:
Good catch. `else` removed

griffin28 committed:
Removed else

biagas commented on src/plots/Volume/VolumeAttributes.code:
```diff
@@ -605,7 +604,6 @@ VolumeViewerEnginePluginInfo::~VolumeViewerEnginePluginInfo()
     }
 }
 
-
```
The removal of these lines caused a problem with the building of our docs, that use a literal include to  code contained in this file (the line numbers used no longer exist). You can revert this file, or modify  doc\dev_manual\UpdatingPlugins.rst line 51.

griffin28 replied:
Updating now...

These literal includes don't seem like a good practice for the docs or should be used very sparingly. We got lucky here that the lines don't exist so it showed up as an error. For the literal include referencing `../../plots/Volume/CMakeLists.txt` there's no error but the lines it's referring to are wrong as there were additions to the `CMakeLists.txt`  file. It's almost impossible to know if some `literalinclude` in the docs needs updating when making changes unless you do get an error in the build. I suspect most cases you'll just have silent errors as the lines its referencing are still valid. I wonder how many of these silent errors are currently present in the docs.

biagas replied:
I agree with your assessment, sorry the practice caused problems with your PR.

biagas commented on src/doc/using_visit/Plots/PlotTypes/VolumePlot.rst:
```diff
@@ -54,6 +54,10 @@ Each sample point along the ray is composited to form a final color for the scre
 Rays are traced from closest to farthest to allow for early ray termination which stops the sampling process when the pixel opacity gets above a certain threshold.
 This method of volume-rendering yields superior pictures at the cost of speed and memory use.
 
+If VisIt_ is built with `ANARI <https://www.khronos.org/api/index_2017/anari/>`_ support by using the ``-–anari`` option, there will be an additional rendering method (:numref:`Figure %s<anari_rendering_atts_window>`). 
```
Building the docs is complaing about line 57, that `anari_rendering_atts_window` label is not defined.

griffin28 replied:
Fixed. Did a `make manuals` and everything looks good.

biagas reviewed (changes requested):
Building the docs is failing. I read the raw log, and have made comments related to the warnings (which are treated as errors in our CI)

griffin28 committed:
Update to docs from PR review
